### PR TITLE
change(web): rotation polish - OSK reload no longer needed ⬅️

### PR DESCRIPTION
**Pull request by jahorton**
Based on #6786, as it addresses related aspects of KMW.  Merging this one first would also make reproducing #6786's underlying issue significantly more difficult.

This should significantly KMW's performance on device rotations, as we no longer throw out OSK DOM structure on device rotation.  It is now reused and adjusted via 15.0's new internal `.refreshLayout()` function chain.

Adjacent to some of the affected code was a note about reinstating something that, well, never got revisited during 15.0 work.  It was relatively simple to reinstate the original purpose of the commented-out function, so I threw that fix in.

# User Testing

## Test environments

- GROUP_IOS - any iOS device should suffice
- GROUP_ANDROID - any Android device should suffice

## Tests

- **TEST_INAPP_ROTATE_P-TO-L**:

  1. With Keyman not active, load it in a portrait orientation.
  2. Once loaded, rotate the device to a landscape orientation.
      - The test "passes" if the OSK rotates properly.

- **TEST_INAPP_ROTATE_L-TO-P**:

  1. With Keyman not active, load it in a landscape orientation.
  2. Once loaded, rotate the device to a portrait orientation.
      - The test "passes" if the OSK rotates properly.

- **TEST_BROWSER_ROTATE_P-TO-L**:

  1. Before loading the test page, have the device in a portrait orientation.
  2. Load the KeymanWeb testing page "Test unminified Keymanweb".
  3. Select the top text area in order to display the OSK.
  4. Open the language menu, but do not change keyboards.
  5. Once loaded, rotate the device to a landscape orientation.
      - The test "passes" if the OSK rotates properly.
      - The test "fails" if the OSK is not visible after rotation.

- **TEST_BROWSER_ROTATE_L-TO-P**:

  1. Before loading the test page, have the device in a landscape orientation.
  2. Load the KeymanWeb testing page "Test unminified Keymanweb".
  3. Select the top text area in order to display the OSK.
  4. Open the language menu, but do not change keyboards.
  5. Once loaded, rotate the device to a portrait orientation.
      - The test "passes" if the OSK rotates properly.
      - The test "fails" if the OSK is not visible after rotation.

### CODE REVIEW

File: web/source/osk/anchoredOskView.ts [hunk 1]
```diff
@@ -59,12 +61,34 @@ namespace com.keyman.osk {
       s.position = 'fixed';
     }
 
-    protected postKeyboardLoad() {
-      // Initializes the size of a touch keyboard.
+    /**
+     * @override
+     */
+    public refreshLayout(pending?: boolean): void {
+      // Prevent a circular call-stack chain.
```
Comment (mcdurdin) [inline web/source/osk/anchoredOskView.ts:68]: Can we explain in this comment how we'd get a recursion -- I think that what `isResizing` is doing is clear enough, but tracing through to figure out how we get recursion and why that is not a bug would be helpful to me.

File: web/source/kmwrotation.ts [hunk 2]
```diff
@@ -45,16 +45,11 @@ namespace com.keyman {
...
-      // TODO:  Reattach later in the refactoring process!
-      //osk.hideLanguageList();
-      // Force a re-layout for the active keyboard's currently-utilized layout.
-      if(this.keyman.core.activeKeyboard) {
-        this.keyman.core.activeKeyboard.refreshLayouts();
-      }
+      osk.hideLanguageMenu();
 
-      osk._Load();
+      osk.setNeedsLayout()
```
Comment (mcdurdin) [inline web/source/kmwrotation.ts:50]:
```suggestion
      osk.setNeedsLayout();
```

File: web/source/kmwembedded.ts [hunk 3]
```diff
@@ -256,10 +256,7 @@ namespace com.keyman.text {
    */
   keymanweb['correctOSKTextSize']=function() {
     let osk: com.keyman.osk.AnchoredOSKView = keymanweb.osk;
-    if(osk?.vkbd) {
-      osk._Load(); // TODO:  replace with osk.refreshLayout() in the future once it can perfectly
-                   //        handle rotations.
-    }
+    osk.refreshLayout();
```
Comment (mcdurdin) [inline web/source/kmwembedded.ts:259]:
```suggestion
    keymanweb.osk?.refreshLayout();
```

The previous code used optional chaining for `osk` -- could `osk` ever be nullish?

Also, we might as well dump the local variable here when this is a one-liner.

File: web/source/osk/oskView.ts [hunk 4]
```diff
@@ -1031,7 +1037,7 @@ namespace com.keyman.osk {
      * Scope        Public
      * @param       {(boolean|number)=}      bShow     True to display, False to hide, omitted to toggle
      */
-     ['show'](bShow: boolean) {
+     ['show'](bShow?: boolean) {
```
Comment (mcdurdin) [inline web/source/osk/oskView.ts:1040]: This bothers me -- why are we toggling visibility of the language menu from `showLanguageMenu()`?

File: web/source/osk/anchoredOskView.ts (hunk 1 above)
Comment (jahorton) [reply web/source/osk/anchoredOskView.ts:68]: image

https://github.com/keymanapp/keyman/blob/c66bfac43f63c829d6b7d226029f26951cd49021/web/source/osk/anchoredOskView.ts#L82-L87

https://github.com/keymanapp/keyman/blob/c66bfac43f63c829d6b7d226029f26951cd49021/web/source/osk/anchoredOskView.ts#L138-L145

For the line 441 call...

https://github.com/keymanapp/keyman/blob/c66bfac43f63c829d6b7d226029f26951cd49021/web/source/osk/oskView.ts#L404-L404

https://github.com/keymanapp/keyman/blob/c66bfac43f63c829d6b7d226029f26951cd49021/web/source/osk/oskView.ts#L441-L442

And you can probably follow it from there, given the nomenclature.

Comment (jahorton) [reply web/source/osk/anchoredOskView.ts:68]: Note that `setSize` can also be used on OSK init and is called by different subclasses of OSKView.

For example, the desktop-style `FloatingOSKView` calls it when the user manually resizes the OSK during a click & drag on the resize handle.

File: web/source/kmwembedded.ts (hunk 3 above)
Comment (jahorton) [reply web/source/kmwembedded.ts:259]: Note that the 'optional chaining' used by 'the previous code' was based on whether or not the `VisualKeyboard` component of the OSK had been initialized.  Not whether or not the main `OSKView` itself was nullish.

File: web/source/osk/oskView.ts (hunk 4 above)
Comment (jahorton) [reply web/source/osk/oskView.ts:1040]: I... uh, this isn't the relevant spot to put that comment, so I'm a little confused as to what you're actually asking.  Unless I'm missing something?  But the contents of the tagged method don't refer to the language menu whatsoever, so I'm at a loss.

If you're referring to our toggle in the rotation manager:

https://github.com/keymanapp/keyman/blob/c66bfac43f63c829d6b7d226029f26951cd49021/web/source/kmwrotation.ts#L44-L50

That's simply pre-existing behavior.  Or, at least, it was, and we're now restoring it.

Comment (mcdurdin) [reply web/source/osk/oskView.ts:1040]: Apologies. This comment should have gone on line 989:

https://github.com/keymanapp/keyman/blob/a328a38297f3b2fec42bf3a08d14cbd44dcb64a9/web/source/osk/oskView.ts#L986-L991

But I see that it is on a new menu, so it would always be showing it anyway. Eliminating the unintuitive toggle behaviour altogether for a missing parameter would be better ... but as it is a public API we probably can't do that.

File: web/source/kmwembedded.ts (hunk 3 above)
Comment (mcdurdin) [reply web/source/kmwembedded.ts]: The old code did a `osk?.`:

```
  if(osk?.vkbd) {
```

Which would be `false` if `osk` is nullish. So from what I can see, it did test whether or not the main `OSKView` itself was nullish?

File: web/source/osk/anchoredOskView.ts (hunk 1 above)
Comment (mcdurdin) [reply web/source/osk/anchoredOskView.ts:68]: Thanks. Could you summarize in the code comment?

File: web/source/osk/oskView.ts (hunk 4 above)
Comment (jahorton) [reply web/source/osk/oskView.ts:1040]: Well... for the language menu, it's been like this for a _while_.

https://github.com/keymanapp/keyman/blob/86965b57408706eef6126ff2846c7799044ef7c4/web/source/kmwosk.ts#L1723-L1729

This was part of #1496.

That said, note how that `show` function is different from the one noted (originally) above.  The main OSK itself has been documented and published with that toggle behavior... so... yeah.

**Review by mcdurdin — APPROVED** [review]

**Comment by bharanidharanj** [comment]
### GROUP_ANDROID: - any Android device should suffice

* **TEST_INAPP_ROTATE_P-TO-L (PASSED):** Tested in the Android 11.0 Mobile device and it is working as expected. 
* **TEST_INAPP_ROTATE_L-TO-P (PASSED):** Tested in the Android 11.0 Mobile device and it is working as expected.
* **TEST_BROWSER_ROTATE_P-TO-L (FAILED):** Tested in the Android 11.0 Mobile device along with the corresponding web page and I noticed that the OSK was gone after rotating the device from Portrait to Landscape mode. 
 


* **TEST_BROWSER_ROTATE_L-TO-P (FAILED):** Tested in the Android 11.0 Mobile device along with the corresponding web page and I noticed that the OSK was gone after rotating the device from Landscape to Portrait mode.

**Comment by jahorton** [comment]
> ### GROUP_ANDROID: - any Android device should suffice
>     * **TEST_BROWSER_ROTATE_P-TO-L (FAILED):** Tested in the Android 11.0 Mobile device along with the corresponding web page and I noticed that the OSK was gone after rotating the device from Portrait to Landscape mode.
>     * **TEST_BROWSER_ROTATE_L-TO-P (FAILED):** Tested in the Android 11.0 Mobile device along with the corresponding web page and I noticed that the OSK was gone after rotating the device from Landscape to Portrait mode.

Good catches; looks like I missed something subtle with that.  I've just made a tweak that should fix that behavior.  Fortunately, it's in browser-only code, so we don't need to toss the app-based test results; a retest on just those two failed tests ought be fine.

GROUP_ANDROID TEST_BROWSER_ROTATE_P-TO-L: OPEN
GROUP_ANDROID TEST_BROWSER_ROTATE_L-TO-P: OPEN

**Comment by MakaraSok** [comment]
# Test Results

### GROUP_IOS: - any iOS device should suffice

* **TEST_INAPP_ROTATE_P-TO-L (PASSED):** the keyboard rotates as usual.
* **TEST_INAPP_ROTATE_L-TO-P (PASSED):** the keyboard rotates as usual.
* **TEST_BROWSER_ROTATE_P-TO-L (PASSED):** the keyboard rotates as usual.
* **TEST_BROWSER_ROTATE_L-TO-P (FAILED):** the keyboard has about one fifth of the right part cropped off

Landscape:
IMG_5BDB15DFB369-1


Rotate to portraite:
IMG_768EE7FDB198-1

**Comment by jahorton** [comment]
> # Test Results
> * **TEST_BROWSER_ROTATE_L-TO-P (FAILED):** the keyboard has about one fifth of the right part cropped off

Interesting... this does not reproduce in Chrome mobile-device emulation.  Looks like I'll have to try something closer to a real device.

**Comment by bharanidharanj** [comment]
### GROUP_ANDROID: - any Android device should suffice

* **TEST_BROWSER_ROTATE_P-TO-L (PASSED):** Able to see the Keyboard after the P-TO-L rotation. 

From Portrait


To Landscape



* **TEST_BROWSER_ROTATE_L-TO-P (FAILED):** Keyboard is not visible after L-TO-P rotation.
From Landscape 


To Portrait

**Comment by mcdurdin** [comment]
> Interesting... this does not reproduce in Chrome mobile-device emulation. Looks like I'll have to try something closer to a real device.

Smells like a race condition / order of operations -- if so it may not consistently repro also...

**Comment by jahorton** [comment]
> Smells like a race condition / order of operations -- if so it may not consistently repro also....

Well, sort-of / kind-of.  Has to do with how odd `window.innerWidth` acts in mobile device browsers.

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 16.0.43-alpha](https://keyman.com/downloads/releases/alpha/16.0.43)
